File: 141-linked-list-cycle/141-linked-list-cycle.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
       ListNode lagging=head;
       ListNode leading=head;
       while(leading!=null&&leading.next !=null){
           leading=leading.next.next;
           lagging=lagging.next;
           if(leading==lagging){
               return true;
           }
       }
        return false;
    }
}

File: 141-linked-list-cycle/141-linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode*lagging=head;
        ListNode*leading=head;
        while(leading!=NULL&&leading->next!=NULL){
            leading=leading->next->next;
            lagging=lagging->next;
            if(lagging==leading){
                return true;
            }
        }
        return false;
    }
};